### ISSUE TITLE
Better auth strategy for react-native / mobile app.

### ISSUE THREAD

**Issue by burtonator** [hicommonwealth/commonwealth#10297]
## Describe the bug

tl;dr I think we need a better auth strategy for the mobile app.  For the time being though I'm using a work around that's less than ideal.

Normally, in the mobile app, if you load a URL, it should open that URL in the browser.

For example, if someone creates a post that links to a URL.  Say for example:

https://www.youtube.com/watch?v=dQw4w9WgXcQ

That should NOT load in our mobile app. Instead, we call Linking.openURL which causes the main browser (Safari, Chrome, etc) to load that URL.

This way the user is using a real browser and the mobile app still functions.

However, during auth, things get a big weird because the oauth flow will trigger loading of external URLs.  Like accounts.google.com which is actually part of the auth flow.  

However, without changing the Linking logic, that opens in the users main browser and they end up authenticated in their BROWSER not the mobile app.

What I *CAN* do as a workaround, in the short term, is to whitelist a bunch of URLs that are used for authentication.

Here is our current whitelist:

```typescript
const ACCEPTED_URL_PATTERNS = [
  /^https:\/\/accounts\.google\.com(\/.*)?$/,
  /^https:\/\/auth\.magic\.link(\/.*)?$/,
  /^https:\/\/appleid\.apple\.com(\/auth\/.*)?$/,
  /^https:\/\/iforgot\.apple\.com(\/password\/.*)?$/,
  /^https:\/\/([a-zA-Z0-9-]+\.)*commonwealth\.im(\/.*)?$/,
  /^https:\/\/([a-zA-Z0-9-]+\.)*common\.xyz(\/.*)?$/,
  /^http:\/\/192\.168\.\d{1,3}\.\d{1,3}(\/.*)?$/,  // Matches http://192.168.x.x with optional path

  // https://discord.com/oauth2/authorize?client_id=1311417374058217535&redirect_uri=http%3A%2F%2F192.168.86.31%3A8080%2Ffinishsociallogin&state=eyJjcnlwdG9DaGFsbGVuZ2VTdGF0ZSI6Ik9BR2JzOHNXcEpCYVdIbUlic1VNLVFqTWdpMGl6UkxoNTdwdlYyaWN1eUxDUndUcVh%2BV29CTDFsUXJmSFJVNDJuWUhkLjAzMHZxdXBTbmRQa3haS0NMS0NkdUl1WUFlNWdDdTFvN2tZVWVlN3pNR1VFQzBhSlV4bGUxY0pJRXJWIn0%3D&scope=identify%20email&code_challenge=PYdB3gjqwtH1lo0EC7M_8uUVXqvwYqFlv_7Q0kkaUXg&code_challenge_method=S256&response_type=code
  // https://discord.com/api/oauth2/authorize?client_id=1311417374058217535&redirect_uri=http%3A%2F%2F192.168.86.31%3A8080%2Ffinishsociallogin&state=eyJjcnlwdG9DaGFsbGVuZ2VTdGF0ZSI6IkZPenpHV1dsV1k5c2drMUZqN3FiZXBuZElPM2pOdGk4SEt1WmptelUyQn5PYm9vMm1ISzN1b0I1QjNQM0RaZnlHX1pLRkdiLjJMRF9XLWVpUTlOeWNtVXJtLmJFaVoxUnV1dzhXSkZpeXo5anZBOW12M1FEdXl6SUpLZGxiWFJBIn0%3D&scope=identify%20email&code_challenge=YN47v3U7Yx7gBpVHVdJrB97gYKoSLUzZibvzzn0Ns-Y&code_challenge_method=S256&response_type=code

  /^https:\/\/discord\.com\/(api\/oauth2\/authorize|oauth2\/authorize)(\/.*)?(\?.*)?$/ // Matches both /api/oauth2/authorize and /oauth2/authorize with query parameters
]

```

However, this is non-idea for a number of reasons:

- I have to manually go through and find those URL and verify that each auth provider works
- The URLs could change at any time, breaking our normal auth flow.

... but this is the worst one.

The browser has NO access to your existing used password or cached credentials.  

This makes auth REALLY hard because usually these things are one click in the browser, but now, in our app, they ALWAYs require the password.

Most of my browser passwords at this point are cached or in lastpass.

This is a pretty poor experience for our users.  

It's actually really impacting my ability to debug because I don't know my passwords off the top of my head.

I think there could a fix though.

- On the mobile app we always trigger the browser to perform auth.  
- They validate who they are by a normal authentication flow
- Then we redirect BACK into the mobile app using deep linking 

What I'm worried about though is that if we used a nonce to authenticate the in the mobile browser, this could in theory be intercepted.  

I can't think of a simpler way to do this though.  

One idea could be, for mobile auth, is we send a challenge code to their email, then they enter that in the mobile app manually.  

Then they're in. 

I think any way we go is not going to be super fun because all the solutions I can think of have serious tradeoffs.

@dillchen @Rotorsoft @timolegros @kurtisassad 

### Initial conditions

**Environment:** 

**Branch/Release version:** 

**Browser:** 

**Wallet:**
